### PR TITLE
Configurable timeout for custom operation handler

### PR DESCRIPTION
**Pull request by PradeepKiruvale**
## Proposed changes

The custom operation must timeout if not completed successfully within the given timeout period.
The `timeout` will be passed as part of the operation file, for example as below.

```
[exec]
  topic = "c8y/s/ds"
  on_message = "511"
  command = "/usr/share/tedge-contrib/bin/handle_command.sh"
  timeout = 60
```
As in the above operation file, if the operation does not complete within mentioned timeout then it will be gracefully killed by sending the `sigterm` and will wait for some time.
If the operation is not stopped even after sending the `sigterm` then it will be killed forcefully by sending the `sigkill`.

## Types of changes

- [ ] Bugfix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [x] Improvement (general improvements like code refactoring that doesn't explicitly fix a bug or add any new functionality)
- [ ] Documentation Update (if none of the other choices apply)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)

## Paste Link to the issue
<br/>

## Checklist

- [x] I have read the [CONTRIBUTING](https://github.com/thin-edge/thin-edge.io/blob/main/CONTRIBUTING.md) doc
- [x] I have signed the [CLA](https://github.com/thin-edge/thin-edge.io/blob/main/CONTRIBUTOR-LICENSE-AGREEMENT.md) (in all commits with git commit -s)
- [x] I ran `cargo fmt` as mentioned in [CODING_GUIDELINES](https://github.com/thin-edge/thin-edge.io/blob/main/CODING_GUIDELINES.md)
- [x] I used `cargo clippy` as mentioned in [CODING_GUIDELINES](https://github.com/thin-edge/thin-edge.io/blob/main/CODING_GUIDELINES.md)
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] I have added necessary documentation (if appropriate)

### CODE REVIEW

File: crates/common/logged_command/src/logged_command.rs [hunk 1]
```diff
@@ -16,18 +18,60 @@ pub struct LoggingChild {
 
 impl LoggingChild {
     pub async fn wait_with_output(
-        self,
+        mut self,
         logger: &mut BufWriter<File>,
+        time_out: Option<usize>,
```
Comment (didier-wenzek) [inline crates/common/logged_command/src/logged_command.rs:23]: Pass the timeout as a `Duration` not as some number with unknown  units.

File: crates/common/logged_command/src/logged_command.rs [hunk 2]
```diff
@@ -16,18 +18,60 @@ pub struct LoggingChild {
...
     ) -> Result<Output, std::io::Error> {
+        match time_out {
+            Some(timeout) => {
+                tokio::select! {
+                    _ = self.inner_child.wait() => {
+                        self.get_outcome(logger).await
+                    }
```
Comment (didier-wenzek) [inline crates/common/logged_command/src/logged_command.rs:30]: Something is weird here as we start to await the child with `self.inner_child.wait()` *ignoring* its output, then when terminated we await again, using that time `self.inner_child.wait_with_output()`. Reading the [documentation](https://docs.rs/tokio/latest/tokio/process/struct.Child.html#method.wait_with_output) it's not clear which *remaining* output is collected.

Why not something along these lines:
```suggestion
                    outcome = self.inner_child.wait_with_output() => {
                        self.finalize_outcome(logger, outcome).await
                    }
```

File: crates/common/logged_command/src/logged_command.rs [hunk 3]
```diff
@@ -16,18 +18,60 @@ pub struct LoggingChild {
...
+                    _ = tokio::time::sleep(std::time::Duration::from_secs(timeout.try_into().unwrap())) => {
+                        self.timeout_operation(logger, timeout).await
+                    }
```
Comment (didier-wenzek) [inline crates/common/logged_command/src/logged_command.rs:33]: This is not clear and hides the same issue as the first arm of the `tokio::select`, i.e. calling different `wait` then `wait_with_output` making unclear which part of the output is collected.

**Review by didier-wenzek — COMMENTED** [review]
The whole timeout logic has to be rewritten. But fist, add tests for this new feature.

```
let outcome_fut = self.inner_child.wait_with_output();

// first attempt
tokio::select! {
    outcome  = outcome_fut => {
         return self.finalize(outcome)
    }
   _ = tokio::time::sleep(duration) => {
       send_term
   }
}


// 2nd attempt
tokio::select! {
    outcome  = outcome_fut => {
         return self.finalize(outcome)
    }
   _ = tokio::time::sleep(duration) => {
       send_kill
   }
}

outcome  = outcome_fut.wait();
return  self.finalize(outcome);

```

File: crates/common/logged_command/src/logged_command.rs (hunk 2 above)
Comment (PradeepKiruvale) [reply crates/common/logged_command/src/logged_command.rs:31]: If we do as suggested above I doubt we cover all the below cases
```
match time_out {
 **// If the timeout is provided then get  inside**       
         Some(timeout) => { 
                tokio::select! {
          
         **// if the operation completes before timeout then get the outcome**   
                    _ = self.inner_child.wait() => {                      self.get_outcome(logger).await
                    }
                   **// if the operation did not complete within the given time, then kill it forcefully, collect the outcome, and update the error message accordingly.**
                    _ = tokio::time::sleep(timeout) => {
                        self.timeout_operation(logger).await
                    }
                }
            }
            **// If the timeout is not provided then, wait till it completes then collect the outcome**
            None => self.get_outcome(logger).await,
        }
```

Comment (didier-wenzek) [reply crates/common/logged_command/src/logged_command.rs:31]: The whole `tokio::select!` logic has to be rewritten and tests added to cover the 4 cases (no timeout, the script terminates before the timeout, the script is killed with SIGTERM, the script is killed with SIGKILL).

File: crates/common/logged_command/src/logged_command.rs [hunk 4]
```diff
@@ -2,12 +2,15 @@ use log::error;
...
 use std::process::Output;
 use std::process::Stdio;
+use std::time::Duration;
 use tokio::fs::File;
 use tokio::io::AsyncWriteExt;
 use tokio::io::BufWriter;
 use tokio::process::Child;
 use tokio::process::Command;
 
+const OPERATION_TIMEOUT: Duration = tokio::time::Duration::from_secs(10); // in seconds
```
Comment (didier-wenzek) [inline crates/common/logged_command/src/logged_command.rs:12]: The comment is useless ;-)

File: crates/common/logged_command/src/logged_command.rs (hunk 2 above)
Comment (PradeepKiruvale) [reply crates/common/logged_command/src/logged_command.rs:32]: Added tests to test all the four cases

File: crates/common/logged_command/src/logged_command.rs [hunk 5]
```diff
@@ -16,18 +18,60 @@ pub struct LoggingChild {
...
+            .as_mut()
+            .unwrap()
+            .stderr
+            .append(&mut "operation failed due to timeout".as_bytes().to_vec());
         outcome
     }
 }
 
+async fn send_sig_term(child: &Child, time_out: usize) {
+    let pid = nix::unistd::Pid::from_raw(child.id().unwrap() as nix::libc::pid_t);
```
[the remark this answers is not in the document]
Comment (PradeepKiruvale) [reply crates/common/logged_command/src/logged_command.rs]: fixed

File: crates/common/logged_command/src/logged_command.rs (hunk 4 above)
Comment (PradeepKiruvale) [reply crates/common/logged_command/src/logged_command.rs]: ;)

File: crates/common/logged_command/src/logged_command.rs (hunk 2 above)
Comment (PradeepKiruvale) [reply crates/common/logged_command/src/logged_command.rs:32]: > > Added tests to test all the four cases
> 
> Thank you. However, these tests doesn't cover the point I would like to clarify and the issue I see to start with `inner_child.wait()` and use `inner_child.wait_with_output()` only _after_ a timeout. **Is the output of the command correctly collected?** 

Yes, we are collecting the output correctly. The `inner_child.wait()`, just returns the exit status, when it finishes successfully. The outcome of the `command` (success or killed) has to be collected using the `inner_child.wait_with_output()`. This is what we are doing here.

> The scripts used for these tests must echo some content and the tests must check the content emitted before the script has been killed is as expected.

Yes, I did echo some text for only one test and then asserted on that content as well. I can do the same for all the other tests as well.

> 
> ```
>     let script = r#"#!/usr/bin/bash    
>     for i in {1..50}
>     do
>         echo $i
>         sleep 1
>     done
>     "#;
> ```
> 
> As side note, it might be simpler to write these tests here - at the level of the `logged_command`, rather than behind the mapper.

I thought about it. Then decided to go at the `mapper` level because there were no tests for the `custom` operations at the mapper level. This will cover now both the `logged_command` as well as the mapper level.

Comment (PradeepKiruvale) [reply crates/common/logged_command/src/logged_command.rs:32]: Updated the tests to assert the `execution status and stdout/stderr ouptut` that is logged.

Comment (didier-wenzek) [reply crates/common/logged_command/src/logged_command.rs:32]: > Yes, we are collecting the output correctly. 

I'm still not convinced by the tests as only the latest line is checked.

Comment (PradeepKiruvale) [reply crates/common/logged_command/src/logged_command.rs:32]: Do you want to assert the whole text of the log?

Comment (didier-wenzek) [reply crates/common/logged_command/src/logged_command.rs:32]: The improved version of  `timeout_operation()` is far better now with a second `tokio::select!` to await either a timeout or the child to exit gracefully. However, I would really prefer that a *single* future is created to await the child and to *re-use* the same future in all the timeout checks. The tests are still only not checking the whole content of the child stdout and stderr while it's really not clear what part of these logs will be actually captured when calling `inner_child.wait_with_output()` only after the termination of the child.

Comment (PradeepKiruvale) [reply crates/common/logged_command/src/logged_command.rs:32]: I tried to have a single future, but there is an issue with sharing the child IDs across two different futures. 
The tests are now asserting the whole output of the logs. 

The inner_child_with_output, will capture the entire output of the command execution, it includes the stdout, stderr, and also the exit status.

File: crates/core/tedge_mapper/src/c8y/converter.rs [hunk 6]
```diff
@@ -634,13 +635,15 @@ fn create_supported_operations(path: &Path) -> Result<Message, ConversionError>
 
         Ok(Message::new(
             &Topic::new_unchecked(&stopic),
-            Operations::try_new(path)?.create_smartrest_ops_message()?,
+            Operations::try_new(path, Duration::from_secs(0), Duration::from_secs(0))?
```
Comment (didier-wenzek) [inline crates/core/tedge_mapper/src/c8y/converter.rs:638]: What does it mean to have a timeout of 0 seconds?

File: crates/tests/tedge_test_utils/src/fs.rs [hunk 7]
```diff
@@ -77,11 +78,11 @@ impl TempTedgeDir {
 }
 
 impl TempTedgeFile {
-    pub fn with_raw_content(self, content: &str) {
+    pub fn with_raw_content(&self, content: &str) {
```
Comment (didier-wenzek) [inline crates/tests/tedge_test_utils/src/fs.rs:81]: Why this change?

**Review by didier-wenzek — COMMENTED** [review]
I would really avoid to add new config settings in `tedge.toml`. For many reasons:

- This will have to be re-implemented soon with the new config under implementation.
- This is not urgent. This is for *default* values. We can hard-code them. Users can set timeout  values to each operations if not happy with the defaults.

File: crates/common/tedge_config/src/tedge_config_cli/settings.rs [hunk 8]
```diff
@@ -551,6 +551,31 @@ impl ConfigSetting for FirmwareChildUpdateTimeoutSetting {
     type Value = Seconds;
 }
 
+pub struct CustomOperationGracefulTimeout;
+
+impl ConfigSetting for CustomOperationGracefulTimeout {
+    const KEY: &'static str = "operation.customops.default_graceful_timeout";
```
[the remark this answers is not in the document]
Comment (PradeepKiruvale) [reply crates/common/tedge_config/src/tedge_config_cli/settings.rs:557]: Just for the future,  if there are any configs for the operations, then they can go under `operations`. These timeouts are for `custom operations`.

**Comment by PradeepKiruvale** [comment]
> I would really avoid to add new config settings in `tedge.toml`. For many reasons:
> 
> * This will have to be re-implemented soon with the new config under implementation.
> * This is not urgent. This is for _default_ values. We can hard-code them. Users can set timeout  values to each operations if not happy with the defaults.

The user can set only `graceful_timeout` not the `forceful_timeout`.  Then tests will have to wait for a long for `forceful_timeout`.

File: crates/common/tedge_config/src/tedge_config_cli/settings.rs (hunk 8 above)
[the remark this answers is not in the document]
Comment (didier-wenzek) [reply crates/common/tedge_config/src/tedge_config_cli/settings.rs:557]: A single default timeout is enough for all operations. The users being able to set specific values on specific operations.

[the remark this answers is not in the document]
Comment (didier-wenzek) [reply crates/common/tedge_config/src/tedge_config_cli/settings.rs:557]: But in an y case, I would avoid to add these settings now.

**Comment by didier-wenzek** [comment]
> > I would really avoid to add new config settings in `tedge.toml`. For many reasons:
> > 
> > * This will have to be re-implemented soon with the new config under implementation.
> > * This is not urgent. This is for _default_ values. We can hard-code them. Users can set timeout  values to each operations if not happy with the defaults.
> 
> The user can set only `graceful_timeout` not the `forceful_timeout`. Then tests will have to wait for a long for `forceful_timeout`.

You added [ `OnMessageExec::forceful_timeout`](https://github.com/thin-edge/thin-edge.io/pull/1917/files#diff-0c4d041da038aab99dfe396742f19e353f2091ae727044f9751fdde7574a9336R33), so the users can now set  `forceful_timeout` ;-)

For testing purpose, it makes sense to pass default timeout values when building `OnMessageExec` objects. But these defaults values can be hardcoded by the mapper with no need to read a config file.

**Comment by PradeepKiruvale** [comment]
> > > I would really avoid to add new config settings in `tedge.toml`. For many reasons:
> > > 
> > > * This will have to be re-implemented soon with the new config under implementation.
> > > * This is not urgent. This is for _default_ values. We can hard-code them. Users can set timeout  values to each operations if not happy with the defaults.
> > 
> > 
> > The user can set only `graceful_timeout` not the `forceful_timeout`. Then tests will have to wait for a long for `forceful_timeout`.
> 
> You added [ `OnMessageExec::forceful_timeout`](https://github.com/thin-edge/thin-edge.io/pull/1917/files#diff-0c4d041da038aab99dfe396742f19e353f2091ae727044f9751fdde7574a9336R33), so the users can now set `forceful_timeout` ;-)
> 
> For testing purpose, it makes sense to pass default timeout values when building `OnMessageExec` objects. But these defaults values can be hardcoded by the mapper with no need to read a config file.

Passing can be done in case of unit tests, what about the integration tests?

File: crates/core/tedge_mapper/src/c8y/converter.rs (hunk 6 above)
Comment (PradeepKiruvale) [reply crates/core/tedge_mapper/src/c8y/converter.rs:638]: These tests do not need any timeout options, but the API needs those parameters. I can make them optional, but this will make the code more complicated. So, I went this way.

File: tests/RobotFramework/tests/cumulocity/Operation/c8y_Command_Fails [hunk 9]
```diff
@@ -0,0 +1,5 @@
+[exec]
```
Comment (rina23q) [inline tests/RobotFramework/tests/cumulocity/Operation/c8y_Command_Fails:1]: The content of `c8y_Command_Fails` and `c8y_Command_Success` are the same. Can't you just have one file?

File: crates/extensions/c8y_mapper_ext/src/converter.rs [hunk 10]
```diff
@@ -72,6 +72,7 @@ use tedge_mqtt_ext::Topic;
 use tedge_utils::size_threshold::SizeThreshold;
 use thiserror::Error;
 use time::format_description::well_known::Rfc3339;
+use tokio::time::Duration;
```
Comment (rina23q) [inline crates/extensions/c8y_mapper_ext/src/converter.rs:75]: Not `use std::time::Duration;`?

File: crates/common/logged_command/src/logged_command.rs [hunk 11]
```diff
@@ -28,6 +86,13 @@ impl LoggingChild {
     }
 }
 
+async fn send_sig_term(child: &Child) {
+    if let Some(pid) = child.id() {
```
Comment (rina23q) [inline crates/common/logged_command/src/logged_command.rs:90]: If `child.id()` returns `None`, this function does nothing. I think it should return error as it means no child process is there.

File: crates/common/logged_command/src/logged_command.rs [hunk 12]
```diff
@@ -16,10 +18,66 @@ pub struct LoggingChild {
...
+                _ = tokio::time::sleep(timeout) => {
+                    self.inner_child.kill().await?;
+                }
+            }
+        }
+
+        let mut outcome = self.inner_child.wait_with_output().await;
+
+        // update the stderr message
+        outcome.as_mut().unwrap().stderr.append(
```
Comment (rina23q) [inline crates/common/logged_command/src/logged_command.rs:72]: This `unwrap()` is not safe.

File: tests/RobotFramework/tests/cumulocity/Operation/c8y_Command_Fails (hunk 9 above)
Comment (rina23q) [inline tests/RobotFramework/tests/cumulocity/Operation/c8y_Command_Fails:1]: Also be careful with the directory name. We already have tests/RobotFramework/tests/cumulocity/**operation**. You may rename from `Operation` to `custom_operation`. (just suggestion, you can use another name, but `Operation` is very much conflict.

**Review by rina23q — COMMENTED** [review]
Left some comments.

File: tests/RobotFramework/tests/cumulocity/Operation/c8y_Command_Fails (hunk 9 above)
Comment (PradeepKiruvale) [reply tests/RobotFramework/tests/cumulocity/Operation/c8y_Command_Fails:1]: Yes, make sense. Renamed it.

File: crates/common/logged_command/src/logged_command.rs (hunk 12 above)
Comment (PradeepKiruvale) [reply crates/common/logged_command/src/logged_command.rs:72]: fixed

File: crates/common/logged_command/src/logged_command.rs (hunk 11 above)
Comment (PradeepKiruvale) [reply crates/common/logged_command/src/logged_command.rs:90]: fixed

File: tests/RobotFramework/tests/cumulocity/Operation/c8y_Command_Fails (hunk 9 above)
Comment (PradeepKiruvale) [reply tests/RobotFramework/tests/cumulocity/Operation/c8y_Command_Fails:1]: There is a `forceful_timeout` that was missing. I added it.

File: crates/common/logged_command/src/logged_command.rs (hunk 3 above)
Comment (PradeepKiruvale) [reply crates/common/logged_command/src/logged_command.rs]: not valid anymore

File: crates/core/c8y_api/src/smartrest/operations.rs [hunk 13]
```diff
@@ -10,18 +10,33 @@ use crate::smartrest::smartrest_serializer::SmartRestSetSupportedOperations;
...
 pub struct OnMessageExec {
     command: Option<String>,
     on_message: Option<String>,
     topic: Option<String>,
     user: Option<String>,
+    #[serde(rename = "timeout")]
+    #[serde_as(as = "Option<DurationSeconds<u64>>")]
+    pub graceful_timeout: Option<Duration>,
+    #[serde_as(as = "Option<DurationSeconds<u64>>")]
+    pub forceful_timeout: Option<Duration>,
```
[the remark this answers is not in the document]
Comment (PradeepKiruvale) [reply crates/core/c8y_api/src/smartrest/operations.rs]: Added a function to deserialize it to duration.

File: crates/tests/tedge_test_utils/src/fs.rs (hunk 7 above)
Comment (PradeepKiruvale) [reply crates/tests/tedge_test_utils/src/fs.rs]: reverted

File: crates/common/logged_command/src/logged_command.rs (hunk 11 above)
Comment (didier-wenzek) [reply crates/common/logged_command/src/logged_command.rs]: No the former code was correct. If [`child.id()`](https://docs.rs/tokio/latest/tokio/process/struct.Child.html#method.id) returns `None` this simply means that the child process no more running. This is not an error as our point was to kill it!

File: crates/common/logged_command/src/logged_command.rs [hunk 14]
```diff
@@ -16,10 +19,70 @@ pub struct LoggingChild {
...
+            Ok(v) => {
+                v.stderr.append(
+                    &mut format!(
+                        "operation failed due to timeout: duration={}s",
+                        timeout.as_secs()
+                    )
+                    .as_bytes()
+                    .to_vec(),
+                );
+                Ok(v.clone())
```
Comment (didier-wenzek) [inline crates/common/logged_command/src/logged_command.rs:81]: Why this `clone()` ? It's surprising to clone and drop `v`.

File: crates/common/logged_command/src/logged_command.rs (hunk 2 above)
Comment (didier-wenzek) [reply crates/common/logged_command/src/logged_command.rs:32]: My comment from April 21st is still valid. I don't understand how this can work.

Sorry, but I find the current proposal contrived and not convincing. I cannot blindly approved it.
So, I will take the time to understand how the whole output of the child can be captured while being ignored at first.

File: crates/common/logged_command/src/logged_command.rs (hunk 11 above)
Comment (PradeepKiruvale) [reply crates/common/logged_command/src/logged_command.rs]: Yes, thats right. I reverted it.

File: crates/common/logged_command/src/logged_command.rs (hunk 14 above)
Comment (PradeepKiruvale) [reply crates/common/logged_command/src/logged_command.rs:81]: It was under `mut ref`. To return I had to do that, but I fixed it with the correct solution `to_owned`.

File: crates/extensions/c8y_mapper_ext/src/converter.rs [hunk 15]
```diff
@@ -1529,11 +1541,23 @@ mod tests {
 
         let now = std::time::Instant::now();
         converter
-            .execute_operation("5", "sleep", "sleep_one")
+            .execute_operation(
+                "5",
+                "sleep",
+                "sleep_one",
+                tokio::time::Duration::from_secs(0),
+                tokio::time::Duration::from_secs(0),
```
Comment (didier-wenzek) [inline crates/extensions/c8y_mapper_ext/src/converter.rs:1549]: Using 0 seconds as a timeout makes this test useless.

Indeed the point is to check that the converter returns immediately despite the operations take some time to run. But the timeout force the command to return immediately. So the test might be green but for bad reasons.

File: crates/common/logged_command/src/logged_command.rs [hunk 16]
```diff
@@ -26,6 +54,86 @@ impl LoggingChild {
...
+            );
+            Ok(v)
+        }
+        Err(e) => Err(std::io::Error::new(ErrorKind::Other, e.to_string())),
+    };
+
+    outcome
+}
+
+fn send_sigal_to_stop_child(child: Option<u32>, signal_type: CmdStatus) {
```
Comment (didier-wenzek) [inline crates/common/logged_command/src/logged_command.rs]: Why an `Option` ?

File: crates/common/logged_command/src/logged_command.rs [hunk 17]
```diff
@@ -26,6 +54,86 @@ impl LoggingChild {
...
+                .as_bytes()
+                .to_vec(),
+            );
+            Ok(v)
+        }
+        Err(e) => Err(std::io::Error::new(ErrorKind::Other, e.to_string())),
+    };
+
+    outcome
+}
```
Comment (didier-wenzek) [inline crates/common/logged_command/src/logged_command.rs:122]: No don't hide the error behind a fake io error.

In fact it would be even simpler to take as parameter an input

```suggestion
fn update_stderr_message(
    mut output: Output
    timeout: Duration,
) -> Result<Output, std::io::Error> {
            output.stderr.append(
                &mut format!(
                    "operation failed due to timeout: duration={}s",
                    timeout.as_secs()
                )
                .as_bytes()
                .to_vec(),
            );

    output
}
```

File: crates/common/logged_command/src/logged_command.rs [hunk 18]
```diff
@@ -1,20 +1,48 @@
...
+    ) -> Result<Output, std::io::Error> {
+        let cid = self.inner_child.id();
+        let cmd_line = self.command_line;
+        let mut status = CmdStatus::Successful;
+        tokio::select! {
+            outcome = self.inner_child.wait_with_output() => {
+               Self::update_and_log_outcome(cmd_line, outcome, logger, graceful_timeout, &mut status).await
+            }
+            _ = Self::timeout_operation(&mut status, cid, graceful_timeout, forceful_timeout) => {
+                Err(std::io::Error::new(std::io::ErrorKind::Other,"failed to read the output"))
```
Comment (didier-wenzek) [inline crates/common/logged_command/src/logged_command.rs:41]: One failed to kill the process not to read its output.

```suggestion
                Err(std::io::Error::new(std::io::ErrorKind::Other,"failed to kill the process: {cmd_line}"))
```

**Review by didier-wenzek — COMMENTED** [review]
The main point of this PR is now correct.

Still some small points to be fix. Most importantly the test in c8y mapper converter that is now green for bad reasons.

File: crates/common/logged_command/src/logged_command.rs (hunk 16 above)
Comment (PradeepKiruvale) [reply crates/common/logged_command/src/logged_command.rs]: To send the `signal` if the child is present.

**Review by didier-wenzek — APPROVED** [review]
Approved - Thank you.
